monkey/monkey_island/cc/agent_event_handlers/handle_ping_scan_event.py

```diff
@@ -0,0 +1,60 @@
+from logging import getLogger
+
+from common.agent_events import PingScanEvent
+from monkey_island.cc.models import CommunicationType, Machine
+from monkey_island.cc.repository import (
+    IAgentRepository,
+    IMachineRepository,
+    INodeRepository,
+    RetrievalError,
+    StorageError,
+    UnknownRecordError,
+)
+
+logger = getLogger(__name__)
+
+
+class handle_ping_scan_event:
+    def __init__(
+        self,
+        agent_repository: IAgentRepository,
+        machine_repository: IMachineRepository,
+        node_repository: INodeRepository,
+    ):
+        self._agent_repository = agent_repository
+        self._machine_repository = machine_repository
+        self._node_repository = node_repository
+
+    def __call__(self, event: PingScanEvent):
+        if not event.response_received:
+            return
+
+        try:
+            dest_machine = self._get_destination_machine(event)
+            self._update_destination_machine(dest_machine, event)
+            src_machine = self._get_source_machine(event)
+
+            # Update or create the node
+            self._node_repository.upsert_communication(
+                src_machine.id, dest_machine.id, CommunicationType.SCANNED
+            )
+        except (RetrievalError, StorageError, TypeError, UnknownRecordError):
+            logger.exception("Unable to process ping scan data")
+
+    def _get_destination_machine(self, event: PingScanEvent) -> Machine:
+        try:
+            dest_machines = self._machine_repository.get_machines_by_ip(event.target)
+            return dest_machines[0]
+        except UnknownRecordError:
+            machine = Machine(id=self._machine_repository.get_new_id())
+            self._machine_repository.upsert_machine(machine)
+            return machine
+
+    def _get_source_machine(self, event: PingScanEvent) -> Machine:
+        agent = self._agent_repository.get_agent_by_id(event.source)
+        return self._machine_repository.get_machine_by_id(agent.machine_id)
+
+    def _update_destination_machine(self, machine: Machine, event: PingScanEvent):
+        if event.os is not None:
```

Comment:
```suggestion
        if event.os is not None and machine.operating_system is None:
```

ICMP packets are a very coarse way of determining OS. It's possible an agent has already run on the machine and has more accurate information. This component can't know whether it's findings are more or less accurate than what's already there. The safe thing to do is just assume what's already in the machine is more accurate.